### PR TITLE
[gas] remove test_only and optional, and support versioned gas schedule

### PR DESCRIPTION
**Pull request by vgao1996**
This makes a few improvements to the macros we use to generate the bindings between the Rust representation of the gas schedule and its on-chain form. Specifically, this
1. Removes `test_only`. `test_only` doubles the complexity of the macros and all it does is to allow us to assign non-zero gas costs to two test-only natives in move-stdlib, which isn't really necessary.
2. Replaces the `optional` marker with a new syntax that allows one to define a versioned gas schedule.

Here's how the versioned gas schedule work:
```rust
/* before */
"foo" // or
optional "foo"

/* after */
"foo" // means the entry is expected to be found in all versions, under key "foo"
{ 1.. => "foo" } // means the entry is expected to be found in version 1 and above, under key "foo"
{ 3..=4 => "foo" } // means the entry is expected to be found in version 3 and 4 under key "foo", but not above 3 or below 4
{ 1 => "foo", 2.. => "bar" } // means the entry is expected to be found under key "foo" in version, and under key "bar" in version 2 and above
```
This new system effectively allows one to rename or delete gas parameters.

### CODE REVIEW

File: api/src/context.rs [hunk 1]
```diff
@@ -912,14 +912,15 @@ impl Context {
 
             let gas_schedule_params =
                 match GasScheduleV2::fetch_config(&storage_adapter).and_then(|gas_schedule| {
+                    let featrure_version = gas_schedule.feature_version;
```
Comment (movekevin) [inline api/src/context.rs:915]: nit: typo featrure_version => feature_version

Comment (movekevin) [reply api/src/context.rs:915]: Also nit: Why not just inline this if this var is only used once?

**Review by movekevin — APPROVED** [review]

File: aptos-move/aptos-gas/src/aptos_framework.rs [hunk 2]
```diff
@@ -72,26 +72,26 @@ crate::natives::define_gas_parameters_for_natives!(GasParameters, "aptos_framewo
...
     [.type_info.type_name.base, "type_info.type_name.base", 300 * MUL],
     // TODO(Gas): the on-chain name is wrong...
     [.type_info.type_name.per_byte_in_str, "type_info.type_name.per_abstract_memory_unit", 5 * MUL],
-    [.type_info.chain_id.base, optional "type_info.chain_id.base", 150 * MUL],
+    [.type_info.chain_id.base, { 4.. => "type_info.chain_id.base" }, 150 * MUL],
 
     // Reusing SHA2-512's cost from Ristretto
-    [.hash.sha2_512.base, optional "hash.sha2_512.base", 3_240],
-    [.hash.sha2_512.per_byte, optional "hash.sha2_512.per_byte", 60],
+    [.hash.sha2_512.base, { 4.. => "hash.sha2_512.base" }, 3_240],
```
[the remark this answers is not in the document]
Comment (vgao1996) [reply aptos-move/aptos-gas/src/aptos_framework.rs:87]: 4 or above. It's basically a rust pattern

File: api/src/context.rs (hunk 1 above)
Comment (vgao1996) [reply api/src/context.rs:915]: > nit: typo featrure_version => feature_version

Thanks for catching this! I can't believe I made so many typos yesterday...

> Also nit: Why not just inline this if this var is only used once?

Tried that. Borrow checker's not happy.

Comment (davidiw) [inline api/src/context.rs:915]:
```suggestion
                    let feature_version = gas_schedule.feature_version;
```

File: api/src/context.rs [hunk 3]
```diff
@@ -912,14 +912,15 @@ impl Context {
...
                     let gas_schedule = gas_schedule.to_btree_map();
-                    AptosGasParameters::from_on_chain_gas_schedule(&gas_schedule)
+                    AptosGasParameters::from_on_chain_gas_schedule(&gas_schedule, featrure_version)
```
Comment (davidiw) [inline api/src/context.rs:917]:
```suggestion
                    AptosGasParameters::from_on_chain_gas_schedule(&gas_schedule, feature_version)
```

**Review by davidiw — APPROVED** [review]

**Pull request merged**